File: Java/Behavioral/MementoPattern.java
package Behavioral;

class Memento {
    private String state;

    public Memento(String state) {
        this.state = state;
    }

    public String getState() {
        return state;
    }
}

class Originator {
    private String state;

    public void setState(String state) {
        this.state = state;
    }

    public String getState() {
        return state;
    }

    public Memento saveStateToMemento() {
        return new Memento(state);
    }

    public void getStateFromMemento(Memento memento) {
        state = memento.getState();
    }
}

class Caretaker {
    private Memento memento;

    public void saveState(Memento memento) {
        this.memento = memento;
    }

    public Memento restoreState() {
        return memento;
    }
}

public class MementoPattern {
    public static void main(String[] args) {
        Originator originator = new Originator();
        Caretaker caretaker = new Caretaker();

        originator.setState("State1");
        caretaker.saveState(originator.saveStateToMemento());

        originator.setState("State2");
        System.out.println("Current State: " + originator.getState());

        originator.getStateFromMemento(caretaker.restoreState());
        System.out.println("Restored State: " + originator.getState());
    }
}

File: Python/Behavioral/MementoPattern.py
class Memento:
    def __init__(self, state):
        self.state = state

class Originator:
    def __init__(self):
        self.state = None

    def set_state(self, state):
        self.state = state

    def get_state(self):
        return self.state

    def save_to_memento(self):
        return Memento(self.state)

    def restore_from_memento(self, memento):
        self.state = memento.state

# Usage
originator = Originator()
originator.set_state("State 1")

caretaker = Memento("Saved State")
caretaker.restore_from_memento()

originator.set_state("State 2")
print(f"Current State: {originator.get_state()}")  # Output: Current State: State 2

originator.restore_from_memento(caretaker)
print(f"Restored State: {originator.get_state()}")  # Output: Restored State: State 1
